### PR TITLE
Email template design for general alerts emails

### PR DESCRIPTION
**Pull request by qifeng-bai**
Some improvements on Biosecurity

### CODE REVIEW

File: grails-app/conf/logback.xml [hunk 1]
```diff
@@ -16,7 +16,7 @@
     <logger name="org.quartz" level="ERROR" />
     <logger name="au.org.ala.bootstrap3" level="INFO" />
 
-    <logger name="au.org.ala" level="DEBUG" />
+    <logger name="au.org.ala" level="INFO" />
```
Comment (nickdos) [inline grails-app/conf/logback.xml:19]: This is not used for deploying to prod/test, as the Ansible script pushes [a different copy](https://github.com/AtlasOfLivingAustralia/ala-install/blob/master/ansible/roles/alerts/templates/logback.xml) of this file onto those servers. So might be fine to leave as DEBUG for dev purposes?

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 2]
```diff
@@ -0,0 +1,93 @@
+package au.org.ala.alerts
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:1]: Missing ALA copyright text at top

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import com.jayway.jsonpath.JsonPath
+import com.jayway.jsonpath.PathNotFoundException
+import grails.converters.JSON
+import org.grails.web.json.JSONArray
+import org.grails.web.json.JSONObject
+
+import java.text.SimpleDateFormat
+
+
+class AnnotationService {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:12]: Missing Javadoc for class

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    def httpService
+    def sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss")  // Adjust the pattern as needed
+
+    String appendAssertions(Query query, JSONObject occurrences) {
+        String baseUrl = query.baseUrl
+
+        if(occurrences.occurrences) {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:19]: inconsistent white space after `if`

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import org.grails.web.json.JSONObject
+
+import java.text.SimpleDateFormat
+
+
+class AnnotationService {
+    def httpService
+    def sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss")  // Adjust the pattern as needed
+
+    String appendAssertions(Query query, JSONObject occurrences) {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:16]: You added javadoc on next method (`diff()`), so may as well be consistent and include it on this method, too.

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+     * @param recordJsonPath
+     * @return
+     */
+    def diff(previous, last, recordJsonPath) {
+        // uuid -> occurrence record map
+        def oldRecordsMap = [:]
+        def curRecordsMap = [:]
+        try {
+            oldRecordsMap = JsonPath.read(previous, recordJsonPath).collectEntries { [(it.uuid): it] }
+        }catch (PathNotFoundException e){
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:65]: whitespace before `catch`

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+     * for my annotation alerts, same occurrence record could exist in both result but have different assertions.
+     * so comparing occurrence uuid is not enough, we need to compare 50001/50002/50003 sections inside each occurrence record
+
+     * return a list of records that their annotations have been changed or deleted
+     * @param previous
+     * @param last
+     * @param recordJsonPath
+     * @return
+     */
+    def diff(previous, last, recordJsonPath) {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:59]: Service methods should have typed args, I think. Should have defined return type too.

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        def oldRecordsMap = [:]
+        def curRecordsMap = [:]
+        try {
+            oldRecordsMap = JsonPath.read(previous, recordJsonPath).collectEntries { [(it.uuid): it] }
+        }catch (PathNotFoundException e){
+            log.warn("Previous result is empty or doesn't have any records containing a field ${recordJsonPath} defined in recordJsonPath")
+        }
+        try {
+             curRecordsMap = JsonPath.read(last, recordJsonPath).collectEntries { [(it.uuid): it] }
+        }catch (PathNotFoundException e){
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:70]: whitespace

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        // if an occurrence record doesn't exist in previous result (added) or has different open_assertions or verified_assertions or corrected_assertions than previous (changed).
+        def records = curRecordsMap.findAll {
+            def record = it.value
+            def previousRecord = oldRecordsMap.get(record.uuid)
+            if (previousRecord) {
+                String currentAssertions = JSON.stringify(filterAssertions(record.user_assertions))
+                String previousAssertions = JSON.stringify(filterAssertions(previousRecord.user_assertions))
+                currentAssertions || previousAssertions
+            } else {
+                true
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/AnnotationService.groovy:82]: Be consistent with `return` syntax - either don't include the `return` keyword everywhere or do include it everywhere. Mixing it up makes code harder to read.

File: grails-app/services/au/org/ala/alerts/DatasetService.groovy [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package au.org.ala.alerts
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DatasetService.groovy:1]: ALA copyright text

File: grails-app/services/au/org/ala/alerts/DatasetService.groovy [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+class DatasetService {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DatasetService.groovy:3]: Javadoc

File: grails-app/services/au/org/ala/alerts/DiffService.groovy [hunk 12]
```diff
@@ -40,6 +48,55 @@ class DiffService {
...
+    /**
+     * Indicates if the result of a query has changed by checking its properties.
+     *
+     * todo: when fireWhenChange is true, it only compare the last 'ID', UUID, UID etc field of the previous and current result
+     * The ID values may be same, but the rest of the records may have changed.
+     *
+     * @param queryResult
+     * @return
+     */
+
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DiffService.groovy:60]: Don't need extra line here

File: grails-app/services/au/org/ala/alerts/DiffService.groovy [hunk 13]
```diff
@@ -40,6 +48,55 @@ class DiffService {
...
+        }
+
+        //Example, in a blog/news query,fireWhenNotZero and fireWhenChange both are false
+
+        if (queryService.checkChangeByDiff(queryResult.query)) {
+            log.debug("[QUERY " + queryResult.query.id + "] Has change check. Checking JSON for query : " + queryResult.query.name)
+            changed = hasChangedJsonDiff(queryResult)
+        }
+
+        changed
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DiffService.groovy:97]: Be consistent with `return` keyword - other methods use it, so I suggest always include the `return`, as that is what the rest of the (alerts) code base is already doing.

File: grails-app/services/au/org/ala/alerts/DiffService.groovy [hunk 14]
```diff
@@ -141,48 +199,88 @@ class DiffService {
...
-                        // if an occurrence record exists in previous result but not in current (deleted).
-                        records.addAll(oldRecordsMap.findAll { !curRecordsMap.containsKey(it.value.uuid) }.values())
+                        records = myAnnotationService.diff(previous, last, queryResult.query.recordJsonPath)
+                    } else if (queryService.isAnnotation(queryResult.query)) {
+                        records = annotationService.diff(previous, last, queryResult.query.recordJsonPath)
+                    } else if (queryService.isDatasetQuery(queryResult.query)) {
+                        records = datasetService.diff(queryResult)
+                    } else if (queryService.isDatasetResource(queryResult.query)) {
+                        records = dataResourceService.diff(queryResult)
+                    }else if ( queryService.isBiocacheImages(queryResult.query)) {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DiffService.groovy:213]: whitespace

File: grails-app/services/au/org/ala/alerts/DiffService.groovy [hunk 15]
```diff
@@ -141,48 +199,88 @@ class DiffService {
...
                 }
             } catch (Exception ex) {
-                log.error "queryId: " + queryResult.query.id + ", JsonPath error: ${ex}"
+                log.error("queryId: ${queryResult.query.id}, JsonPath error: ${ex}")
             }
         }
         records
     }
 
+    def findNewRecordsById(QueryResult qs) {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DiffService.groovy:230]: You added Javadoc for the next method you added (`findNewRecordsById` - line 256), so may as well add it here too.

File: grails-app/services/au/org/ala/alerts/DiffService.groovy [hunk 16]
```diff
@@ -141,48 +199,88 @@ class DiffService {
...
+     */
+    def findNewRecordsById(previous, last, recordJsonPath, idJsonPath) {
+        def records = []
+        String fullRecordJsonPath = recordJsonPath + "." + idJsonPath
+        List<String> ids1 = []
+        List<String> ids2 = []
+        try {
+            ids1 = JsonPath.read(last, fullRecordJsonPath)
+            ids2 = JsonPath.read(previous, fullRecordJsonPath)
+        }catch (PathNotFoundException e){
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DiffService.groovy:264]: whitespace

File: grails-app/services/au/org/ala/alerts/DiffService.groovy [hunk 17]
```diff
@@ -141,48 +199,88 @@ class DiffService {
...
+            log.info("it's not an error. Result doesn't contain any records since the returned json does not contain any 'recordJsonPath', if result is empty.")
+        }
+
+        List<String> diff = ids1.findAll { !ids2.contains(it) }
+        //pull together the records that have been added
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DiffService.groovy:269]: comment is above an empty line - so unclear if it applies to line above or below - convention is to put it directly above the line it refers to.

File: grails-app/services/au/org/ala/alerts/DiffService.groovy [hunk 18]
```diff
@@ -141,48 +199,88 @@ class DiffService {
...
+        allRecords.each { record ->
+            if (diff.contains(record.get(idJsonPath))) {
+                records.add(record)
+            }
+        }
+        int maxRecords = grailsApplication.config.getProperty("biosecurity.query.maxRecords", Integer, 500)
+        if (records.size() > maxRecords) {
+            records = records.subList(0, maxRecords)
+        }
+        return records
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DiffService.groovy:281]: Consistent whitespace - other methods have a blank line before the final return statement.

File: grails-app/services/au/org/ala/alerts/HttpService.groovy [hunk 19]
```diff
@@ -0,0 +1,100 @@
+package au.org.ala.alerts
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/HttpService.groovy:1]: ALA copyright text

File: grails-app/services/au/org/ala/alerts/HttpService.groovy [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+import grails.converters.JSON
+import org.apache.commons.io.IOUtils
+import groovy.json.JsonSlurper
+
+
+class HttpService {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/HttpService.groovy:8]: Javadoc

File: grails-app/conf/logback.xml (hunk 1 above)
Comment (qifeng-bai) [reply grails-app/conf/logback.xml:19]: It is used for local dev environment. In most cases, we only need INFO, but sometime, it requires DEBUG

File: grails-app/services/au/org/ala/alerts/HttpService.groovy [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    def grailsApplication
+
+    /**
+     *
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/HttpService.groovy:12]: Needs a Javadoc title/text - method name of `post` is too generic to impart any meaning to the reader.

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy (hunk 2 above)
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/AnnotationService.groovy:1]: Added

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy (hunk 3 above)
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/AnnotationService.groovy:12]: added

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy (hunk 5 above)
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/AnnotationService.groovy:16]: Done

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy (hunk 4 above)
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/AnnotationService.groovy:19]: Fixed

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy (hunk 7 above)
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/AnnotationService.groovy:59]: Added

File: grails-app/services/au/org/ala/alerts/AnnotationService.groovy (hunk 8 above)
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/AnnotationService.groovy:70]: fixed

File: grails-app/services/au/org/ala/alerts/DataResourceService.groovy [hunk 22]
```diff
@@ -0,0 +1,26 @@
+package au.org.ala.alerts
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/DataResourceService.groovy:1]: copyright and javadoc

File: grails-app/services/au/org/ala/alerts/ImageService.groovy [hunk 23]
```diff
@@ -0,0 +1,25 @@
+package au.org.ala.alerts
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/ImageService.groovy:1]: Ditto

File: grails-app/services/au/org/ala/alerts/MyAnnotationService.groovy [hunk 24]
```diff
@@ -0,0 +1,124 @@
+package au.org.ala.alerts
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/MyAnnotationService.groovy:1]: Ditto

File: grails-app/services/au/org/ala/alerts/MyAnnotationService.groovy [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+            occurrences.put('occurrences', reconstructedOccurrences)
+        }
+
+        return occurrences.toString()
+    }
+
+
+    //Search the assertions which the USER has made
+    //return those have been open-issued, verified or corrected
+    //
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/MyAnnotationService.groovy:99]: delete empty comment line

File: grails-app/services/au/org/ala/alerts/CollectionService.groovy [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+ *   rights and limitations under the License.
+ *
+ */
+
+package au.org.ala.alerts
+
+import groovy.json.JsonOutput
+import groovy.json.JsonSlurper
+
+class CollectionService {
```
[the remark this answers is not in the document]
Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/CollectionService.groovy:20]: added

File: grails-app/services/au/org/ala/alerts/NotificationService.groovy [hunk 27]
```diff
@@ -425,6 +389,8 @@ class NotificationService {
     }
 
     /**
+     * todo: this method is not used by a deprecated method: checkStatusDontUpdate
```
Comment (nickdos) [inline grails-app/services/au/org/ala/alerts/NotificationService.groovy:392]: Did you mean "not used"? Comment makes no sense to me, as it stands.

Comment (qifeng-bai) [reply grails-app/services/au/org/ala/alerts/NotificationService.groovy:392]: LOL,  **IS** used by a deprecated method

**Review by nickdos — APPROVED** [review]

**Pull request merged**